mm2src/coins/utxo/utxo_tx_history_v2.rs

```diff
@@ -0,0 +1,603 @@
+
+#[async_trait]
+impl<Coin, Storage> State for OnIoErrorCooldown<Coin, Storage>
+where
+    Coin: UtxoTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = UtxoTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, _ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        Timer::sleep(30.).await;
+        Self::change_state(FetchingTxHashes::new())
+    }
+}
+
+// States have to be generic over storage type because BchAndSlpHistoryCtx is generic over it
+struct WaitForHistoryUpdateTrigger<Coin, Storage> {
+    phantom: std::marker::PhantomData<(Coin, Storage)>,
+}
+
+impl<Coin, Storage> WaitForHistoryUpdateTrigger<Coin, Storage> {
+    fn new() -> Self {
+        WaitForHistoryUpdateTrigger {
+            phantom: Default::default(),
+        }
+    }
+}
+
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>>
+    for WaitForHistoryUpdateTrigger<Coin, Storage>
+{
+}
+
+#[async_trait]
+impl<Coin, Storage> State for WaitForHistoryUpdateTrigger<Coin, Storage>
+where
+    Coin: UtxoTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = UtxoTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        let wallet_id = ctx.coin.history_wallet_id();
+        loop {
+            Timer::sleep(30.).await;
+            match ctx.storage.history_contains_unconfirmed_txes(&wallet_id).await {
+                Ok(contains) => {
+                    if contains {
+                        return Self::change_state(FetchingTxHashes::new());
+                    }
+                },
+                Err(e) => return Self::change_state(Stopped::storage_error(e)),
+            }
+
+            let current_balances = match ctx.coin.get_addresses_balances().await {
+                Ok(balances) => balances,
+                Err(e) => {
+                    error!("Error {e:?} on balance fetching for the coin {}", ctx.coin.ticker());
+                    continue;
+                },
+            };
+
+            let mut to_update = false;
+            for (address, current_balance) in current_balances {
+                match ctx.balances.entry(address) {
+                    // Do nothing if the balance hasn't been changed.
+                    Entry::Occupied(entry) if *entry.get() == current_balance => {},
+                    Entry::Occupied(mut entry) => {
+                        to_update = true;
+                        entry.insert(current_balance);
+                    },
+                    Entry::Vacant(entry) => {
+                        to_update = true;
+                        entry.insert(current_balance);
+                    },
+                }
+            }
+
+            if to_update {
+                return Self::change_state(FetchingTxHashes::new());
```

Comment:
In `FetchingTxHashes` state, it will fetch history for all addresses - can be quite an overhead. We need only addresses whose balance was changed here.

Reply:
Great advice, will do!

Reply:
This led to a huge refactoring of some of the TX history storage methods and the TX history states.
I tried to cover most of the cases in `tx_history_v2_tests.rs` module.